### ISSUE TITLE
How to hide certain fields in Pydantic models?

### ISSUE THREAD

**Issue by enilsen16** [tiangolo/fastapi#1378]
Example:

I have created_at and updated_at fields that I want to auto populate on the model using default_factory, but I don't want the user to be able to set them via an endpoint. I thought this could be achieved by the read only Field arg but then I can never set it.

What's the best way to go about this?

**Comment by includeamin** [comment]
> Example:
> 
> I have created_at and updated_at fields that I want to auto populate on the model using default_factory, but I don't want the user to be able to set them via an endpoint. I thought this could be achieved by the read only Field arg but then I can never set it.
> 
> What's the best way to go about this?


``` python
from pydantic import BaseModel,Field
from datetime import datetime



class User(BaseModel):
    name:str
    create_at: datetime = datetime.now()

# Option 1
class RenderedUser(User):
    create_at: datetime = Field(default=None,alias="_datetime")

# Option 2
class RenderedUser2(BaseModel):
    name:str


user_in_system = User(name='includeamin')

rendered = RenderedUser(**user_in_system.dict())
rendered_2 = RenderedUser2(**user_in_system.dict())

  

print(rendered)
# output name='includeamin' create_at=None

print(rendered_2)
# output name='includeamin'
```

I'm not sure this is the best way or not.

**Comment by includeamin** [comment]
``` python
from pydantic import BaseModel,Field
from datetime import datetime



class User(BaseModel):
    name:str
    create_at: datetime = datetime.now()

user_in_system = User(name='includeamin')
print(user_in_system.dict(exclude={'create_at'}))
{'name': 'includeamin'}
```

**Comment by enilsen16** [comment]
Sorry I should have clarified, how do you hide them in the openapi spec? :)

**Comment by phy25** [comment]
I mean you could create two models, one internal and one external, but you have to be careful between converting between two models (performance v.s. validation).

**Comment by enilsen16** [comment]
Yeah it doesn't seem like there is a good way to convert between two models without validation?

**Comment by enilsen16** [comment]
> Sorry I should have clarified, how do you hide them in the openapi spec? :)

I did notice that Pydantic has a schema_extra function that you can provide but fast didn't seem to be picking it up when generating the openapi docs.

**Comment by tiangolo** [comment]
Check the docs here: https://fastapi.tiangolo.com/tutorial/response-model/#add-an-output-model

Having a model declared as having an attribute and then making it _not_ have the attribute in some other way would make your editor, linter (and you) believe that _that_ model has _that_ attribute while it doesn't. So it's better/easier to declare the 2 models that inherit from the same base.

**Comment by serain** [comment]
@enilsen16 did you ever find a clean way to hide a particular field from the schema?

In my case I have some required headers that are set downstream by my ingress. FastAPI expects to see those headers but it's confusing for my users to see them in the docs.

**Comment by aberrier** [comment]
Pydantic allows you to define extra arguments on Field.
You can use this at your advantage like so
```python
from pydantic import BaseModel, Field


class CustomBaseModel(BaseModel):
    def dict(self, **kwargs):
        hidden_fields = set(
            attribute_name
            for attribute_name, model_field in self.__fields__.items()
            if model_field.field_info.extra.get("hidden") is True
        )
        kwargs.setdefault("exclude", hidden_fields)
        return super().dict(**kwargs)


class Book(CustomBaseModel):
    author: str
    title: str
    genre: str = Field(..., hidden=True)


m = Book(
    author="Antoine de Saint-Exupéry",
    title="The Little Prince",
    genre="Novella",
)
print(m.genre)  # Novella
print(m.dict())  # {'author': 'Antoine de Saint-Exupéry', 'title': 'The Little Prince'}
```

**Comment by trallnag** [comment]
Does anyone know what method / function FastAPI is using to generate the OpenAPI schema from Pydantic models? We could then remove the to be hidden fields on the fly like in @aberrier's example

**Comment by sgiroux** [comment]
You need to modify the schema processing of Pydantic to achieve this.  Easiest way I've found was to monkey the field_schema function and raise a SkipField error.

You can define your 'hidden' fields as follows:
``` 
from pydantic import BaseModel

class Model(BaseModel): 
    exposed_id: int = Field(1)
    hidden_id: int = Field(1, hidden_from_schema=True)`

```

Apply the patch
```
from pydantic import schema
def field_schema(field: ModelField, **kwargs: Any) -> Any:
    if field.field_info.extra.get("hidden_from_schema", False):
        raise schema.SkipField(f"{field.name} field is being hidden")
    else:
        return original_field_schema(field, **kwargs)

original_field_schema = schema.field_schema
schema.field_schema = field_schema
```

**Comment by botsman** [comment]
@sgiroux thanks!  
I've used your solution in my project and it worked good!

I have some private fields which I want to provide for my "trusted" customers. Obviously, I don't want other people to see such fields in the documentation (because they will not get these fields in their responses anyway).

**Comment by Su57** [comment]
i have the same problem, but i can't find a elegant solution
let's say i have a schema used to create a user:

```python
class UserCreationSchema(BaseModel):
    username: str
    password: str
    created_by: Optional[str] = None
```

but i don't want the `created_by`  field exposed  at the auto-genarated docs, in fact,  the value of it should be generated by the backend.

the official doc mentioned above： [add an output model](https://fastapi.tiangolo.com/tutorial/response-model/#add-an-output-model)  just can't help, it's not the same thing.

so what's the best way to go about this?

**Comment by tonythree** [comment]
Hey @Su57 , I had exactly the same issue. Creating extra models (classes) can help.
In fact this is what FastAPI docs [recommend](https://fastapi.tiangolo.com/tutorial/extra-models/) (they use pydantic).
In your case,
```
class UserBase(BaseModel):
    username: str
    password: str

class UserCreationSchema(UserBase):
    created_by: Optional[str] = None
```
So the autogenerated docs can use UserBase class, while your backend can use UserCreationSchema class.
Hope this helps!

**Comment by gautamvij94** [comment]
Is there a way to conditionally remove fields tagged with `(hidden=True)` based on state in request ? 
Eg. 
`If request.state.external=True` then don't return certain fields defined in the response models as `hidden=True`

**Comment by Leon0824** [comment]
I am using a stupid way:

```py
class UserDB(BaseModel):
    name: str
    password: str

class User(BaseModel):
    name: str

@app.get('/user/{name}', response_model=User)
def create_user(name: str):
    user_db = db.select('user', 'name', name)
    user = User(name=user_db.name)
    return user
```

Just do not inhererit from its brother model.

**Comment by arkanmgerges** [comment]
Maybe this will help, by using `SecretStr` with `Field` and `exclude=True`
```python
from typing import Union

from fastapi import FastAPI
from pydantic import BaseModel, EmailStr, SecretStr

app = FastAPI()
@app.post("/user/", response_model=UserOut)
async def create_user(user: UserIn):
    return user

class UserIn(BaseModel):
    username: str
    password: str
    email: EmailStr
    full_name: Union[str, None] = None

class UserOut(UserIn):
    password: SecretStr = Field(..., exclude=True)
```

![image](https://user-images.githubusercontent.com/7836805/185139644-88cf71d4-2418-493b-ab90-dcfa46ef69a0.png)

**Comment by Xenthys** [comment]
I'm late to the party, but if you want to hide Pydantic fields from the OpenAPI schema definition without either adding underscores (annoying when paired with SQLAlchemy), overriding the `schema.field_schema` function that will display warnings in you logs, you can [customize the schema according to Pydantic's documentation](https://docs.pydantic.dev/usage/schema/#schema-customization).

You can therefore add a schema_extra static method in your class configuration to look for a `hidden` boolean field option, and remove it while still retaining all the features you need.

Here is how I did it:
```py
from pydantic import BaseModel, Field

class User(BaseModel):
    public_field: str
    hidden_field: str = Field(hidden=True)

    class Config:
        @staticmethod
        def schema_extra(schema: dict, model):
            props = {}
            for k, v in schema.get('properties', {}).items():
                if not v.get("hidden", False):
                    props[k] = v
            schema["properties"] = props
```

This can also be done by creating a "proxy" BaseModel class per example.
All your existing classes will therefore obtain this feature without editing them:
```py
from pydantic import BaseModel as _BM

class BaseModel(_BM):
    class Config:
        @staticmethod
        def schema_extra(schema: dict, model):
            props = {}
            for k, v in schema.get('properties', {}).items():
                if not v.get("hidden", False):
                    props[k] = v
            schema["properties"] = props
```

You can certainly optimize this function, but that's the cleanest way I've been able to find.
